File: src/main/java/eu/karelhovorka/zpevnik/text/SongText.java
package eu.karelhovorka.zpevnik.text;


import java.util.HashMap;
import java.util.List;
import java.util.Map;

import eu.karelhovorka.zpevnik.util.SongSettingsFormatter;

import static eu.karelhovorka.zpevnik.util.Preconditions.checkNotNull;

public class SongText {

    private final String originalText;

    private final List<Section> sections;

    public final String title;

    private final SongDisplaySettings displaySettings;

    public SongText(String originalText, String title, SongDisplaySettings songDisplaySettings) {
        checkNotNull(originalText, "originalText");
        checkNotNull(title, "title");
        checkNotNull(songDisplaySettings, "songDisplaySettings");
        this.title = title;
        this.displaySettings = songDisplaySettings;
        this.originalText = originalText;
        if (SectionTokenizer.isValid(originalText)) {
            SectionTokenizer sectionTokenizer = new SectionTokenizer();
            List<Section> tmpList = null;
            try {
                tmpList = sectionTokenizer.getSections(originalText);
            } catch (Exception e) {
                e.printStackTrace();
            }
            this.sections = tmpList;
        } else {
            this.sections = null;
        }
    }


    public static SongText fromRawText(String originalText, String title, SongDisplaySettings songDisplaySettings) {
        String modifiedText = SongSettingsFormatter.modifyTextBasedOnSettings(originalText, songDisplaySettings);
        return new SongText(modifiedText, title, songDisplaySettings);
    }

    public String getOriginalText() {
        return originalText;
    }

    public List<Section> getSections() {
        return sections;
    }

    public String getCss() {
        StringBuilder sb = new StringBuilder();
        if (sections != null && sections.size() >= 0) {
            sb.append(" has-sections");
        }
        sb.append(" " + displaySettings.getCss());
        return sb.toString();
    }

    public Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();
        map.put("title", title);
        map.put("originalText", originalText);
        return map;
    }

    @Override
    public String toString() {
        return "SongText{" +
                "sections=" + sections +
                ", title='" + title + '\'' +
                ", displaySettings=" + displaySettings +
                '}';
    }
}

File: src/main/kotlin/eu/karelhovorka/zpevnik/text/SongText.kt
package eu.karelhovorka.zpevnik.text


import eu.karelhovorka.zpevnik.util.Preconditions.checkNotNull
import eu.karelhovorka.zpevnik.util.SongSettingsFormatter


class SongText(val originalText: String, val title: String, private val displaySettings: SongDisplaySettings) {

    val sections: List<Section>

    val css: String
        get() {
            val sb = StringBuilder()
            sb.append(" has-sections")
            if (sections.size == 1 && sections.first().type == SectionType.UNKNOWN) {
                sb.append(" section-single-unknown")
            }
            sb.append(" " + displaySettings.css)
            return sb.toString()
        }

    init {
        checkNotNull(originalText, "originalText")
        checkNotNull(title, "title")
        checkNotNull(displaySettings, "songDisplaySettings")
        val sectionTokenizer = SectionTokenizer()
        this.sections = sectionTokenizer.getSectionsOrSingleSection(originalText)
    }

    fun toMap(): Map<String, Any> {
        val map = HashMap<String, Any>()
        map.put("title", title)
        map.put("originalText", originalText)
        return map
    }

    override fun toString(): String {
        return "SongText{" +
                "sections=" + sections +
                ", title='" + title + '\'' +
                ", displaySettings=" + displaySettings +
                '}'
    }

    companion object {


        fun fromRawText(originalText: String, title: String, songDisplaySettings: SongDisplaySettings): SongText {
            checkNotNull(originalText, "originalText")
            checkNotNull(title, "title")
            checkNotNull(songDisplaySettings, "songDisplaySettings")
            val modifiedText = SongSettingsFormatter.modifyTextBasedOnSettings(originalText, songDisplaySettings)
            return SongText(modifiedText, title, songDisplaySettings)
        }
    }
}
